src/python/pants/backend/python/tasks/python_repl.py

```diff
@@ -50,13 +54,24 @@ def setup_repl_session(self, targets):
     pex_info.entry_point = entry_point
     return self.create_pex(pex_info)
 
+  def _run_repl(self, pex, **pex_run_kwargs):
+    env = pex_run_kwargs.pop('env', os.environ).copy()
+    pex.run(env=env, **pex_run_kwargs)
+
   # N.B. **pex_run_kwargs is used by tests only.
   def launch_repl(self, pex, **pex_run_kwargs):
-    # While the repl subprocess is synchronously spawned, we rely on process group
-    # signalling for a SIGINT to reach the repl subprocess directly - and want to
-    # do nothing in response on the parent side.
-    def ignore_control_c(signum, frame): pass
-
-    with signal_handler_as(signal.SIGINT, ignore_control_c):
-      env = pex_run_kwargs.pop('env', os.environ).copy()
-      pex.run(env=env, **pex_run_kwargs)
+    running_under_pantsd = self.context.options.for_global_scope().enable_pantsd
+
+    if not running_under_pantsd:
+      # While the repl subprocess is synchronously spawned, we rely on process group
+      # signalling for a SIGINT to reach the repl subprocess directly - and want to
+      # do nothing in response on the parent side.
+      with ExceptionSink.trapped_signals(PythonReplSignalHandler()):
+        self._run_repl(pex, **pex_run_kwargs)
+    else:
+      # In pantsd, this task will be running in a non-main thread,
+      # so we can't override signal handling here.
+      # That said, this means that under pantsd,
+      # Ctrl-C will simply crash the repl (and the daemon).
+      # TODO(#7623) Potential more robust (but more invasive) fix.
+      self._run_repl(pex, **pex_run_kwargs)
```

Comment:
This is fixed as of #7623, we can just replace the whole body of this method with:
```suggestion
      with ExceptionSink.ignoring_signint():
        self._run_repl(pex, **pex_run_kwargs)
```